File: 프로그래머스/Level2/위클리챌린지/5주차_모음사전.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

char arr[5] = {'A', 'E', 'I', 'O', 'U'};
vector<string> total;

void dfs(vector<char> v, int cnt){
    if(v.size()==cnt){
        string tmp = "";
        for(int i=0; i<v.size(); i++){
            tmp = tmp + v[i];
        }
        total.push_back(tmp);
        return ;
    }
    
    for(int i=0; i<5; i++){
        v.push_back(arr[i]);
        dfs(v, cnt);
        v.pop_back();
    } 
}

int solution(string word) {
    int answer = 0;
    vector<char> v;
    total.clear();
    
    for(int i=1; i<=5; i++){
        v.clear();
        dfs(v, i);
    }
    
    sort(total.begin(), total.end());
    
    for(int i=0; i<total.size(); i++){
        if(total[i]==word){
            answer = i+1;
            break;
        }
    }
    
    return answer;
}

File: Source Code/c++/프로그래머스/Level2/위클리챌린지/5주차_모음사전.py
def dfs(my_list, tmp_list, word_dic, visited, cnt, target_sz) :
    
    if cnt == target_sz :
        w = ''
        for tmp in tmp_list:
            w = w + tmp
        my_list.append(w)   
        return
    
    for i in range(0, 5) :
        tmp_list.append(word_dic[i])
        dfs(my_list, tmp_list, word_dic, visited, cnt+1, target_sz)
        tmp_list.pop()
    
    return

def solution(word):
    answer = 0
    my_list = []
    tmp_list = []
    word_dic = ['A', 'E', 'I', 'O', 'U']
    visited = [0 for _ in range(5)]
    
    for i in range(1, 6) :
        dfs(my_list, tmp_list, word_dic, visited, 0, i)
    
    my_list = sorted(my_list)
    
    for i in range(0,len(my_list)) :
        if word == my_list[i] :
            answer = i+1
            break
    
    return answer